File: commit_defender/exit_resolver.py
"""Resolve the final exit code from a Report."""

from .config import Config
from .models import Report


class ExitCodeResolver:
    def __init__(self, config: Config) -> None:
        self.config = config

    def resolve(self, report: Report) -> int:
        """Return 0 (pass) or 1 (block)."""
        # API / infrastructure errors only block when ai_review is a hard gate.
        # When ai_review.blocking=false the review is advisory — a network hiccup
        # must not prevent a commit from landing.
        if report.review.is_error and self.config.ai_review.blocking:
            return 1

        # Lint findings at or above the configured threshold always block.
        blocking_findings = report.findings_at_or_above(self.config.blocking_severity)
        if blocking_findings:
            return 1

        # AI review blocks only when both the config gate and the AI say so.
        if self.config.ai_review.blocking and report.review.blocking:
            return 1

        return 0

File: vscode-extension/src/exitResolver.ts
/**
 * Resolve the final exit code for a hook run.
 *
 * Without rule-based linting in the extension, the only blocking signals are:
 *   1. Any P3 (Critical) AI comment.
 *   2. The model's own `blocking: true` flag.
 *
 * AI errors do NOT block — a network hiccup must not prevent a commit from
 * landing. The user can always `git commit --no-verify` for emergencies.
 */

import { AnalysisReport } from './types.js';
import { reviewStatus } from './reviewOutcome.js';

export type EnforcementPolicy = 'legacy-hook' | 'advisory';

export function resolveExitCode(report: AnalysisReport, policy: EnforcementPolicy = 'legacy-hook'): 0 | 1 {
  // New automatic review uses advisory explicitly; existing hooks keep their policy.
  // A zero enforcement code does not prove that the review completed.
  if (policy === 'advisory') return 0;
  const status = reviewStatus(report.review);
  if (status === 'failed' || status === 'cancelled') { return 0; }
  if (report.review.file_comments.some(c => c.priority === 'P3')) { return 1; }
  if (report.review.blocking) { return 1; }
  return 0;
}
